### PR TITLE
feat: dictionary stub implementation

### PR DESCRIPTION
**Pull request by malandis**
Stub out the dictionary API and document it; also implement port of response types. This covers:

- `dictionary_fetch`
- `dictionary_get_field`
- `dictionary_get_fields`
- `dictionary_increment`
- `dictionary_set_field`
- `dictionary_set_fields`
- `dictionary_remove_field`
- `dictionary_remove_fields`

API, documentation, and response types will be subject to change on implementation.

### CODE REVIEW

File: src/momento/responses/dictionary_data/dictionary_get_field.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    See `SimpleCacheClient` for how to work with responses.
+    """
+
+
+class CacheDictionaryGetField(ABC):
+    """Groups all `CacheDictionaryGetFieldResponse` derived types under a common namespace."""
+
+    @dataclass
+    class Hit(CacheDictionaryGetFieldResponse, ValueStringMixin):
```
Comment (schwern) [inline src/momento/responses/dictionary_data/dictionary_get_field.py:25]: Needs field_string and field_bytes

File: src/momento/responses/dictionary_data/dictionary_get_fields.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            """The items for the dictionary fields, as a mapping from bytes to bytes.
+
+            Returns:
+                TDictionaryBytesBytes
+            """
+            return {
+                field: response.value_bytes
+                for field, response in zip(self.fields, self.responses)
+                if isinstance(response, CacheDictionaryGetField.Hit)
+            }
```
Comment (schwern) [inline src/momento/responses/dictionary_data/dictionary_get_fields.py:51]: The GetField response should return both fields and values so...

```suggestion
            return {
                response.field_bytes: response.value_bytes
                for response in self.responses
                if isinstance(response, CacheDictionaryGetField.Hit)
            }
```

And so on for the rest of the value methods.

File: src/momento/simple_cache_client_async.py [hunk 3]
```diff
@@ -249,6 +268,153 @@ async def delete(self, cache_name: str, key: TScalarKey) -> CacheDeleteResponse:
...
+            ttl (CollectionTtl, optional): TTL for the dictionary in cache.
+                This TTL takes precedence over the TTL used when initializing a cache client.
+                Defaults to CollectionTtl.from_cache_ttl().
+
+        Returns:
+            CacheDictionarySetFieldsResponse: result of the set fields operation.
+        """
+        pass
+
+    async def dictionary_remove_field(
```
Comment (schwern) [inline src/momento/simple_cache_client_async.py:385]: I didn't notice `dictionary_set_fields` and `dictionary_set_field` because this is out of order.

**Review by schwern — COMMENTED** [review]
Looks like a good start. The list PR has a shared behavior to test collection TTLs that might be useful to copy.

**Review by cprice404 — APPROVED** [review]
🙈 I'm happy when @schwern is happy

File: src/momento/responses/dictionary_data/dictionary_fetch.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
+from abc import ABC
+from dataclasses import dataclass
+
+from momento.errors import SdkException
+from momento.typing import TDictionaryBytesBytes, TDictionaryStrBytes, TDictionaryStrStr
```
[the remark this answers is not in the document]
Comment (malandis) [reply src/momento/responses/dictionary_data/dictionary_fetch.py:5]: What was forgotten?

File: src/momento/responses/dictionary_data/dictionary_get_fields.py [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    @dataclass
+    class Hit(CacheDictionaryGetFieldsResponse, ValueStringMixin):
+        """Contains the result of a cache hit."""
+
+        responses: List[CacheDictionaryGetFieldResponse]
+        """The value returned from the cache for the specified field. Use the
+        `value_string` property to access the value as a string."""
+
+        fields: List[bytes]
```
[the remark this answers is not in the document]
Comment (malandis) [reply src/momento/responses/dictionary_data/dictionary_get_fields.py:38]: I'll get to that when I do the implementation for sure

File: src/momento/responses/dictionary_data/dictionary_fetch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (schwern) [reply src/momento/responses/dictionary_data/dictionary_fetch.py:5]: TDictionaryBytesStr was forgotten. There should be 4 value methods. str/str, bytes/bytes, str/bytes, bytes/str.

File: src/momento/simple_cache_client_async.py (hunk 3 above)
Comment (malandis) [reply src/momento/simple_cache_client_async.py]: Changed in [27d8d70](https://github.com/momentohq/client-sdk-python/pull/229/commits/27d8d7038ec90ab87902dece3d14bd0a9fe0ec10)

File: src/momento/responses/dictionary_data/dictionary_get_fields.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (schwern) [reply src/momento/responses/dictionary_data/dictionary_get_fields.py:38]: To be clear, I think we should remove the fields property and just provide dicts.

File: src/momento/responses/dictionary_data/dictionary_fetch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (malandis) [reply src/momento/responses/dictionary_data/dictionary_fetch.py:5]: It wasn't forgotten -- we made the decision to not implement it unless someone asked for it

[the remark this answers is not in the document]
Comment (schwern) [reply src/momento/responses/dictionary_data/dictionary_fetch.py:5]: It's a trivial method to complete the interface. Does it make sense to go back an add it everywhere later when someone inevitably requests it?

File: src/momento/responses/dictionary_data/dictionary_get_field.py (hunk 1 above)
Comment (malandis) [reply src/momento/responses/dictionary_data/dictionary_get_field.py:25]: Fixed in [d06b436](https://github.com/momentohq/client-sdk-python/pull/229/commits/d06b436a2af92f0e88087d840574cadc367d0c23)

File: src/momento/responses/dictionary_data/dictionary_get_fields.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (malandis) [reply src/momento/responses/dictionary_data/dictionary_get_fields.py]: Now uses field from `CacheDictionaryGetField.Hit` in [d8cca31](https://github.com/momentohq/client-sdk-python/pull/229/commits/d8cca3139038f8b766f3f03f5d9d261b734559b1)

File: src/momento/typing.py [hunk 6]
```diff
@@ -10,6 +10,14 @@
 TCollectionName = str
 
 # Dictionary Types
+TDictionaryName = TCollectionName
+TDictionaryField = Union[str, bytes]
+TDictionaryValue = Union[str, bytes]
+TDictionaryFields = Sequence[TDictionaryField]
```
Comment (schwern) [inline src/momento/typing.py:16]: We're going with Iterator for listy inputs, correct?

```suggestion
TDictionaryFields = Iterator[TDictionaryField]
```

File: src/momento/responses/dictionary_data/dictionary_get_fields.py (hunk 2 above)
Comment (malandis) [reply src/momento/responses/dictionary_data/dictionary_get_fields.py]: Fixed as above in [d8cca31](https://github.com/momentohq/client-sdk-python/pull/229/commits/d8cca3139038f8b766f3f03f5d9d261b734559b1)

File: src/momento/responses/dictionary_data/dictionary_fetch.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (malandis) [reply src/momento/responses/dictionary_data/dictionary_fetch.py:5]: Added in [b56e7b4](https://github.com/momentohq/client-sdk-python/pull/229/commits/b56e7b4f4c82d3f4f4748258be9c27a14191eef3)

File: src/momento/typing.py (hunk 6 above)
Comment (malandis) [reply src/momento/typing.py:16]: Yes thanks

Comment (malandis) [reply src/momento/typing.py]: Fixed in [65ab746](https://github.com/momentohq/client-sdk-python/pull/229/commits/65ab746868843ba16c8b99aeb76b1d3c5728fa0e)

**Review by schwern — APPROVED** [review]

**Pull request merged**